pkg/controllers/management/etcdbackup/etcdbackup.go

```diff
@@ -0,0 +1,277 @@
+	"fmt"
+	"time"
+
+	minio "github.com/minio/minio-go"
+	"github.com/minio/minio-go/pkg/credentials"
+	"github.com/rancher/kontainer-engine/drivers/rke"
+	"github.com/rancher/kontainer-engine/service"
+	"github.com/rancher/rancher/pkg/controllers/management/clusterprovisioner"
+	"github.com/rancher/rancher/pkg/rkedialerfactory"
+	"github.com/rancher/rancher/pkg/ticker"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	clusterBackupCheckInterval time.Duration = 5 * time.Minute
+)
+
+type Controller struct {
+	ctx                   context.Context
+	clusterClient         v3.ClusterInterface
+	clusterLister         v3.ClusterLister
+	backupClient          v3.EtcdBackupInterface
+	backupLister          v3.EtcdBackupLister
+	backupDriver          service.EngineService
+	KontainerDriverLister v3.KontainerDriverLister
+}
+
+func Register(ctx context.Context, management *config.ManagementContext) {
+	c := &Controller{
+		ctx:                   ctx,
+		clusterClient:         management.Management.Clusters(""),
+		clusterLister:         management.Management.Clusters("").Controller().Lister(),
+		backupClient:          management.Management.EtcdBackups(""),
+		backupLister:          management.Management.EtcdBackups("").Controller().Lister(),
+		backupDriver:          service.NewEngineService(clusterprovisioner.NewPersistentStore(management.Core.Namespaces(""), management.Core)),
+		KontainerDriverLister: management.Management.KontainerDrivers("").Controller().Lister(),
+	}
+
+	local := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+	}
+	docker := &rkedialerfactory.RKEDialerFactory{
+		Factory: management.Dialer,
+		Docker:  true,
+	}
+	driver := service.Drivers[service.RancherKubernetesEngineDriverName]
+	rkeDriver := driver.(*rke.Driver)
+	rkeDriver.DockerDialer = docker.Build
+	rkeDriver.LocalDialer = local.Build
+	rkeDriver.WrapTransportFactory = docker.WrapTransport
+
+	c.backupClient.AddLifecycle(ctx, "etcdbackup-controller", c)
+	go c.clusterBackupSync(ctx, clusterBackupCheckInterval)
+}
+
+func (c *Controller) Create(b *v3.EtcdBackup) (runtime.Object, error) {
+	cluster, err := c.clusterClient.Get(b.ClusterID, metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	v3.BackupConditionCreated.True(b)
+	b.Filename = getBackupFilename(b.Name, cluster)
+	b.BackupConfig = *cluster.Spec.RancherKubernetesEngineConfig.Services.Etcd.BackupConfig
+	b, err = c.backupClient.Update(b)
+	if err != nil {
+		return nil, err
+	}
+
+	kontainerDriver, err := c.KontainerDriverLister.Get("", service.RancherKubernetesEngineDriverName)
+	if err != nil {
+		return nil, err
+	}
+	err = c.backupDriver.ETCDSave(context.Background(), cluster.Name, kontainerDriver, cluster.Spec, b.Name)
```

Comment:
will this function use backup's secretKey? then need to replace with secret value

Reply:
No. This will send the whole cluster spec object to RKE.